x/gamm/twap/README.md

```diff
@@ -1,9 +1,102 @@
-# TWAP
+# TWAP (Time Weighted Average Price)
 
-We maintain TWAP entries for every gamm pool.
+The TWAP package is responsible for being able to serve TWAPs for every AMM pool.
 
-NOTE: Not yet integrated into state machine, this package is a stub.
+A time weighted average price is a function that takes a sequence of `(time, price)` pairs, and returns a price representing an 'average' over the entire time period. The method of averaging can vary from the classic arithmetic mean, (such as geometric mean, harmonic mean), however we currently only implement arithmetic mean.
 
-## Basic architecture notes
+## Arithmetic mean TWAP
 
-We maintain the list of pools altered within in a block
\ No newline at end of file
+Using the arithmetic mean, the TWAP of a sequence `(t_i, p_i)`, from `t_0` to `t_n`, indexed by time in ascending order, is: $\frac{1}{t_n - t_0}\sum_{i=0}^{n-1} p_i (t_{i+1} - t_i)$. (Notice that the latest price `p_n` isn't used, as it has lasted for a time interval of `0` seconds in this range!)
+
+To illustrate with an example, given the sequence: `(0s, $1), (2s, $5), (5s, $1)`, the arithmetic mean TWAP is: $\frac{\$1 * (2s - 0s) + \$5 * (5s - 3s)}{5s - 0s} = \frac{\$10}{5} = \$2$.
+
+## Computation via accumulators method
+
+The prior example for how to compute the TWAP takes linear time in the number of time entries in a range, which is too inefficient. We require TWAP operations to have constant time complexity (in the number of records).
+
+This is achieved by using an accumulator. In the case of an arithmetic Twap, we can maintain an accumulator from `a_n`, representing the numerator of the Twap expression for the interval `t_0...t_n`, namely $a_n = \sum_{i=0}^{n-1} p_i (t_{i+1} - t_i)$. If we maintain such an accumulator for every pool, with `t_0 = pool_creation_time` to `t_n = current_block_time`, we can easily compute the TWAP for any interval. The twap for the time interval of price points `t_i` to `t_j` is then $twap = \frac{a_j - a_i}{t_j - t_i}$, which is constant time given the accumulator values.
```

Comment:
> Notice that the latest price `p_n` isn't used, as it has lasted for a time interval of `0` seconds in this range

> If we maintain such an accumulator for every pool, with `t_0 = pool_creation_time` to `t_n = current_block_time`, we can easily compute the TWAP for any interval

Would accumulator at `t_n` be equal to `t_{n-1}` given the notice?

Reply:
No, because the accumulator at `t_n` (`a_n`) contains the result of `p_{n-1}`, wheres the accumulator at `t_{n-1}`, (`a_{n-1}`) does not have the result of multiplying by `p_{n-1}`